### PR TITLE
IGNITE-21001 Java Thin Client: Fixed stale client cluster group data after cluster restart.

### PR DESCRIPTION
**Pull request by petrov-mg**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterGroupImpl.java [hunk 1]
```diff
@@ -75,21 +75,22 @@ class ClientClusterGroupImpl implements ClientClusterGroup {
...
+    private ClientClusterGroupImpl(
+        ReliableChannel ch,
+        ClientUtils utils,
+        ProjectionFilters projectionFilters
+    ) {
         this.ch = ch;
         this.utils = utils;
         this.projectionFilters = projectionFilters;
+
+        ch.addChannelFailListener(this::invalidateCachedData);
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterGroupImpl.java:93]: There is memory leak. Every time cluster group is created - reference to this group is stored to the channel and this group can't be garbage collected until client closed.

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ClusterGroupClusterRestartTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    /** */
+    private void assertContainsNodes(ClientClusterGroup grp, int... nodeIdxs) {
+        assertTrue(grp.nodes().containsAll(Arrays.stream(nodeIdxs).mapToObj(idx -> grid(idx).localNode()).collect(Collectors.toList())));
+
+        for (int idx : nodeIdxs)
+            assertNotNull(grp.node(grid(idx).localNode().id()));
+    }
+
+    /** */
+    private void assertNotContainsNodes(ClientClusterGroup grp, Collection<ClusterNode> nodes) {
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/client/thin/ClusterGroupClusterRestartTest.java:76]: Not used

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterGroupImpl.java [hunk 3]
```diff
@@ -62,6 +62,9 @@ class ClientClusterGroupImpl implements ClientClusterGroup {
     /** Projection filters. */
     private final ProjectionFilters projectionFilters;
 
+    /** Client channel from which the cluster group nodes data was previously received. */
+    private volatile ClientChannel topDataSrc;
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterGroupImpl.java:66]: volatile is redundant - always modified under synchronized block

**Review by alex-plekhanov — APPROVED** [review]

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ClusterGroupClusterRestartTest.java (hunk 2 above)
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/client/thin/ClusterGroupClusterRestartTest.java]: Fixed.

**Pull request merged**